crates/subspace-farmer/src/cluster/cache.rs

```diff
@@ -0,0 +1,763 @@
+};
+use crate::farm::{FarmError, PieceCache, PieceCacheOffset};
+use anyhow::anyhow;
+use async_nats::Message;
+use async_trait::async_trait;
+use derive_more::Display;
+use futures::stream::FuturesUnordered;
+use futures::{select, stream, FutureExt, Stream, StreamExt};
+use parity_scale_codec::{Decode, Encode, EncodeLike, Input, Output};
+use std::future::{pending, Future};
+use std::pin::{pin, Pin};
+use std::time::Duration;
+use subspace_core_primitives::{Piece, PieceIndex};
+use tokio::time::MissedTickBehavior;
+use tracing::{debug, error, trace, warn};
+use ulid::Ulid;
+
+/// An ephemeral identifier for a cache
+#[derive(Debug, Copy, Clone, Ord, PartialOrd, Eq, PartialEq, Hash, Display)]
+pub enum ClusterCacheId {
+    Ulid(Ulid),
+}
+
+impl Encode for ClusterCacheId {
+    #[inline]
+    fn size_hint(&self) -> usize {
+        1_usize
+            + match self {
+                ClusterCacheId::Ulid(ulid) => 0_usize.saturating_add(Encode::size_hint(&ulid.0)),
+            }
+    }
+
+    #[inline]
+    fn encode_to<O: Output + ?Sized>(&self, output: &mut O) {
+        match self {
+            ClusterCacheId::Ulid(ulid) => {
+                output.push_byte(0);
+                Encode::encode_to(&ulid.0, output);
+            }
+        }
+    }
+}
+
+impl EncodeLike for ClusterCacheId {}
+
+impl Decode for ClusterCacheId {
+    #[inline]
+    fn decode<I: Input>(input: &mut I) -> Result<Self, parity_scale_codec::Error> {
+        match input
+            .read_byte()
+            .map_err(|e| e.chain("Could not decode `CacheId`, failed to read variant byte"))?
+        {
+            0 => u128::decode(input)
+                .map(|ulid| ClusterCacheId::Ulid(Ulid(ulid)))
+                .map_err(|e| e.chain("Could not decode `CacheId::Ulid.0`")),
+            _ => Err("Could not decode `CacheId`, variant doesn't exist".into()),
+        }
+    }
+}
+
+#[allow(clippy::new_without_default)]
+impl ClusterCacheId {
+    /// Creates new ID
+    #[inline]
+    pub fn new() -> Self {
+        Self::Ulid(Ulid::new())
+    }
+}
+
+/// Broadcast with identification details by caches
+#[derive(Debug, Clone, Encode, Decode)]
+pub struct ClusterCacheIdentifyBroadcast {
+    /// Cache ID
+    pub cache_id: ClusterCacheId,
+    /// Max number of elements in this cache
+    pub max_num_elements: u32,
+}
+
+impl GenericBroadcast for ClusterCacheIdentifyBroadcast {
+    const SUBJECT: &'static str = "subspace.cache.*.identify";
+}
+
+/// Write piece into cache
+#[derive(Debug, Clone, Encode, Decode)]
+struct ClusterCacheWritePieceRequest {
+    offset: PieceCacheOffset,
+    piece_index: PieceIndex,
+    piece: Piece,
+}
+
+impl GenericRequest for ClusterCacheWritePieceRequest {
+    const SUBJECT: &'static str = "subspace.cache.*.write-piece";
+    type Response = Result<(), String>;
+}
+
+/// Read piece index from cache
+#[derive(Debug, Clone, Encode, Decode)]
+struct ClusterCacheReadPieceIndexRequest {
+    offset: PieceCacheOffset,
```

Comment:
As explained during walkthrough both cluster and non-cluster versions are having the same exact API. Cache doesn't store its mapping internally and only does contents scanning on farmer cache's request (see `ClusterCacheContentsRequest`). Since method `read_piece_index` has `offset` as an argument, we have an identical property in this request here.